dialer.go

```diff
@@ -441,3 +435,23 @@ func splitHostPort(s string) (host string, port string) {
 	}
 	return
 }
+
+func runCustomResolver(ctx context.Context, address string, resolver Resolver) (string, error) {
+	inputHost, inputPort := splitHostPort(address)
+	resolved, err := resolver.LookupHost(ctx, inputHost)
+	if err != nil {
+		return "", err
+	}
+
+	// if the resolver doesn't return anything, we'll allow the caller to continue
+	// with the original address
+	if len(resolved) == 0 {
+		return "", nil
+	}
+
+	resolvedHost, resolvedPort := splitHostPort(resolved[0])
+	if resolvedPort == "" {
```

Comment:
Should we check for `inputPort != ""` instead?

Basically having a logic where we pick the port number from the address if it's set, otherwise fallback to the one from the resolver, and otherwise default to 9092?

Reply:
I guess that depends on the priority order we want to use, my thinking here was that we should prefer the resolved port, and if not supplied we'll fall back on the one provided in the original `address`. But yeah, just realizing if neither includes a port number, we'll get into a weird state, so I'll add a test case to cover that.

Reply:
Thinking about user experience, it would be really strange to explicitly specify a port number and have the system connect to something else because the resolver had another port registered. Potentially, this could also be a security concern.